ci/Jenkinsfile

```diff
@@ -1,39 +1,134 @@
+                                sh "touch .env"
+                                sh "cargo test"
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
 
-                echo 'Deploying'
-                sh "sudo /usr/local/bin/permaplant-deploy.sh pr ${WORKSPACE}"
+    parallel_checks["check-frontend"] = {
+        node('docker') {
+            checkout scm
+            def node_image = docker.build("permaplant-node", "./ci/container-images/permaplant-node")
+
+            node_image.inside {
+                checkout scm
+                unstash 'definitions.ts'
+
+                echo "HC SVNT DRACONES"
             }
         }
+    }
+
+    parallel(parallel_checks)
+}
 
-        stage('Deploy Dev') {
-            when {
-                branch "master"
+stage('Build in Docker'){
+    def parallel_build = [:]
+
+    parallel_build['build-backend'] = {
+        node('docker') {
+            checkout scm
+            def rust_image = docker.build("permaplant-rust", "./ci/container-images/permaplant-rust")
+
+            docker.image('postgres:alpine').withRun('-e "POSTGRES_USER=ci" -e "POSTGRES_PASSWORD=ci"') { c ->
+                rust_image.inside("--link ${c.id}:db -e 'DATABASE_URL=postgres://ci:ci@db/ci'") {
+                    checkout scm
+                    sh './ci/build-scripts/build-backend.sh'
+
+                    stash includes: 'backend/target/release/backend', name: 'backend'
+                }
             }
+        }
+    }
+
+    parallel_build['build-frontend'] = {
+        node('docker') {
+            checkout scm
+            def node_image = docker.build("permaplant-node", "./ci/container-images/permaplant-node")
+
+            node_image.inside {
+                checkout scm
+                unstash 'definitions.ts'
 
-            steps {
-                echo 'Deploying to Dev ...'
-                sh "sudo /usr/local/bin/permaplant-deploy.sh dev ${WORKSPACE}"
+                sh './ci/build-scripts/build-frontend.sh'
+                stash includes: 'frontend/dist/**/*', name: 'frontend'
             }
         }
     }
+
+    parallel(parallel_build)
+}
+
+stage('Deploy PR') {
+    node('permaplant') {
+        checkout scm
+
+        echo "Workspace is '${WORKSPACE}'"
+        echo 'Ensure pr service is stopped'
+        sh 'sudo /usr/sbin/service permaplant-pr stop'
+
+        withEnv(['DATABASE_URL=postgres://cidb:cidb@127.0.0.1/cidb']) {
+            echo 'Resetting the CI DB ...'
+            sh 'sudo /usr/local/bin/permaplant-reset-ci.sh'
+
+            echo "Unstash Artifacts"
+            unstash 'schema.rs'
+            unstash 'frontend'
+            unstash 'backend'
+
+            echo 'Deploying'
+            sh "sudo /usr/local/bin/permaplant-deploy.sh pr ${WORKSPACE}"
+        }
+    }
+}
+
+// Deploying to Dev only happens on branch "master"
+if (env.BRANCH_NAME == 'master') {
```

Comment:
Beautiful solution for different deployments!

Reply:
thanks :rocket: